### ISSUE TITLE
PreRelease ToDo: analytical_functions

### ISSUE THREAD

**Issue by nmueller18** [ISAAKiel/mortAAR#20]
I do not understand the new warning messages. I just have generated a life table from the Magdalenenberg data set, which only contains the age range and the number of deceased. I still got the warning "In one of your data.frames are more than the two necessary columns a, Dx. Note that these additional columns will be dropped in the output."

**Comment by nevrome** [comment]
How did you do that? I've also been a bit sloppy here, but we should try to create [reproducible bug reports](https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example).   

If I run the following lines
```
hu <- read.csv("data-raw/magdalenenberg.txt", sep = "\t")

life.table(hu)
```
I get this error: 
```
 Error in function_list[[k]](value) : 
  The data.frames with the following element IDs in the input list don't have the numeric columns 'a' and 'Dx': 1 
```
This is good and expected behavior. What did you do to get the other error?

**Comment by nmueller18** [comment]
Sorry, you are right.

```
td<-magdalenenberg
td <-td %>% replace(td == "60-x", "60-70")
td <- td %>%   tidyr::separate(age, c("from", "to"))
td <- td %>%  transform( from = as.numeric(from), to = as.numeric(to)  )
td_prepared <- prep.life.table( td, dec = "Dx",  agebeg = "from", ageend = "to",   methode = "Standard", age.range = "excluded" )
td_result <- td_prepared %>% life.table()
```

This results in two instead of only one life table and:

```
Warnmeldungen:
1: In FUN(X[[i]], ...) :
  In one of your data.frames are more than the two necessary columns a, Dx. Note that these additional columns will be dropped in the output.
2: In FUN(X[[i]], ...) :
  In one of your data.frames are more than the two necessary columns a, Dx. Note that these additional columns will be dropped in the output.
```

**Comment by nmueller18** [comment]
At the moment, Ax is put between qx and Lx. However, in standard life tables (e. g. UN tables) it is usually placed after ex. Because here we have rel_popx, I would prefer to place it after a. I think I have seen it there in some cases, and from a logical point of view this makes sense to me. Even to somebody who does not read the documentation the connection between a and Ax might become apparent.

**Comment by nevrome** [comment]
So... let's walk through this. 

```
library(magrittr)

td <- magdalenenberg
td <- td %>% replace(td == "60-x", "60-70")

td <- td %>%   tidyr::separate(a, c("from", "to"))
# not: td <- td %>%  tidyr::separate(age, c("from", "to"))

td_prepared <- prep.life.table( 
  td, dec = "Dx",  agebeg = "from", ageend = "to",   methode = "Standard", age.range = "excluded" 
)
```

The result of `prep.life.table()` is this monstrosity:

```
$Bestattungen
   Alter a      Dx
1      0 1  0.9475
2      1 4  2.8425
3      5 5  4.6200
4     10 5  4.5400
5     15 5  4.2100
6     20 5 14.9900
7     25 5 20.6100
8     30 5 17.2000
9     35 5 14.3900
10    40 5  6.6800
11    45 5  4.0400
12    50 5  5.4900
13    55 5  5.7200
14    60 5  2.3600
15    65 5  2.3600

$<NA>
   Alter a Dx
1      0 1  1
2      1 4  4
3      5 5  5
4     10 5  5
5     15 5  5
6     20 5  5
7     25 5  5
8     30 5  5
9     35 5  5
10    40 5  5
11    45 5  5
12    50 5  5
13    55 5  5
14    60 5  5
15    65 5  5
```

Instead of one data.frame we have two in the list td_prepared. And both have the additional column "Alter" which causes the warning in `life.table()` in the first place. 

This is neither a problem of `life.table()` nor of `prep.life.table()`.  `prep.life.table()` is simply not prepared for this kind of input. 

A correct workflow could look like this:

```
td <- magdalenenberg
td <- td %>% replace(td == "60-x", "60-70")
td <- td %>%   tidyr::separate(a, c("from", "to"))
td <- td %>%  transform(from = as.numeric(from), to = as.numeric(to))
td$a <- td$to - td$from
td_prepared <- td[, c("a", "Dx")]
td_result <- td_prepared %>% life.table()
```

The more I think about this, the more I get the impression, that it's a bad idea to provide this data preparation function. It raises the wrong expectation to be the ultimate preparation tool despite it's actually only suitable for a very limited kind of input data. At least we need more documentation about what it exactly does and what not. @whamer

**Comment by nmueller18** [comment]
I do not agree. The function does what it is supposed to do, and it gives the user a better impression what is being done to her data. prep.life.table is important because only there the age range treatment (inclusive vs. exclusive) and the method (actual age ranges) can be specified. In the further process we obviously only have to get rid of "Alter" (which should be renamed anyway) and prevent an additional table of being generated (which is not produced if grnam is not NULL). To me it seems partly to be a problem of the interplay of the functions prep.life.table and life.table.

**Comment by nevrome** [comment]
> To me it seems partly to be a problem of the interplay of the functions prep.life.table and life.table.

Please elaborate. I don't see how `life.table()` is involved here. Is there anything to be changed? 

On one hand I understand your point: Obviously this input function is a great convenience feature. On the other hand there's no better way to avoid the black box you're mentioning here

> ...and it gives the user a better impression what is being done to her data.

but to force the user to do the data wrangling by himself. Your example shows very well, that he has to do some steps alone no matter how comprehensive or sophisticated we manage `prep.life.table()` to be.

Anyway: I guess you will agree, that the function needs more documentation to be useful. At the moment there's only one sentence of description and no example at all.

**Comment by nmueller18** [comment]
> Please elaborate. I don't see how life.table() is involved here. Is there anything to be changed?

life.table() should simply ignore the column "Alter" which is at the moment produced by prep.life.table by default (and should be kept in my opinion for readibility).

> by forcing the user to do the data wrangling by himself. Your example shows very well, that he has to do some steps alone no matter how comprehensive or sophisticated we manage prep.life.table() to be.

Partly you are right. However, after having looked at your vignette and playing around with the data of magdalenenberg and muensingen, I think we should demand that the data is provided by age ranges in the format "#-#". The user has obviously to take care that there are no inconsistencies like "60-x" or "10+/-1". But afterwards, everything should be taken care of by prep.life.table. Especially the lines

`td <- td %>%   tidyr::separate(a, c("from", "to"))`

and

`td <- td %>%  transform(from = as.numeric(from), to = as.numeric(to))`

are always needed.
So, I suggest that prep.life.table kicks in a little earlier than it is the case at the moment. In that case it would require the following arguments:
dec (optional), age(required and in the format "#-#", to be split into agebeg and ageend), grnam (optional), methode (optional with default value), age.range(optional with default value). prep.life.table would then need to have the additional function of separating and transforming (see above). This would have the additional advantage that the user does not have to specify the columns "ageebeg" and "ageend" by hand.
@FFaupel and @whamer : What do you think?

**Comment by nevrome** [comment]
> life.table() should simply ignore the column "Alter" which is at the moment produced by prep.life.table by default (and should be kept in my opinion for readibility).

Hm... [You suggested to throw all additional variables away](https://github.com/ISAAKiel/mortAAR/issues/19#issuecomment-304585914) and we implemented this feature. I even think this was a very good decision. If you want a new "age" column now, I suggest we construct it within `life.table()` from "a".


> However, after having looked at your vignette and playing around with the data of magdalenenberg and muensingen, I think we should demand that the data is provided by age ranges in the format "#-#".

That's a bad idea: Never force a user to create a layout sensitive string if you can avoid it. The solution with the variable "a" is foolproof in comparison. We could create another extra function that creates "a" from a string vector like this.

The input function you're trying to achieve appears to be an "Eierlegende Wollmilchsau" ;-). I'm very skeptical that this is possible.

**Comment by nmueller18** [comment]
Hm .. I get the impression that we do not talk about the same things. I do not want to change anything substantial with `life.table()`. At the moment,` prep.life.table()` simply produces an additional column ("Alter") which irritates` life.table()` that therefore throws the warning message.
Let's talk about the workflow: I think that the user should strongly be recommended to access `life.table()` only after having altered her data via `prep.life.table()` for input. Do we agree on that?

**Comment by nevrome** [comment]
> I think that the user should strongly be recommended to access life.table() only after having altered her data via prep.life.table() for input. Do we agree on that?

As far as I understand it, `prep.life.table()` is a nice and convenient helper - some sort of "quick start" assistant that somehow manages to deal with many use cases. 
Honest question: Why do you say it should be the "recommended" way of using the `life.table()` function? Because of the "methode" and the "age.range" switches?

I suggest we postpone this discussion until `prep.life.table()` is fixed and better documented. It's also more easy to talk about this in person...

**Comment by nmueller18** [comment]
> Honest question: Why do you say it should be the "recommended" way of using the life.table() function? Because of the "methode" and the "age.range" switches?

Yes.

> I suggest we postpone this discussion until prep.life.table() is fixed and better documented. It's also more easy to talk about this in person...

You are right!

**Comment by whamer** [comment]
Hey both of you. What exactly should be changed on the output of the input function. This "monstrosity" was the result of the interaction with the life.table() group, why I thought, that there should be no problems with our output. I have talked to the group and have been told, that the list-form is required as well as the columns "a" and "Dx". Since the function would use the elements of these columns by adressing them directly, the "Alter" column should not be a problem either (as I´ve been told). So what should be fixed at the input function? Should I change the name "Alter" to "age" or should I remove the column?

**Comment by nmueller18** [comment]
I would prefer the following: change "Alter" to "age", and `life.table()` should ignore this column. However, there is still the issue that a second table is generated when "grnam" is not specified. Could you have a look at that?

**Comment by whamer** [comment]
Problem solved. I renamed Alter to Age and the second table is no longer generated if no "grnam" is specified:

> td_prepared <- prep.life.table( 
>     td, dec = "Dx",  agebeg = "from", ageend = "to",   methode = "Standard", age.range = "excluded" 
> )
> 
> 
> td_prepared
>$Deceased
>  Age a      Dx
1    0 1  0.9475
2    1 4  2.8425
3    5 5  4.6200
4   10 5  4.5400
5   15 5  4.2100
6   20 5 14.9900
7   25 5 20.6100
8   30 5 17.2000
9   35 5 14.3900
10  40 5  6.6800
11  45 5  4.0400
12  50 5  5.4900
13  55 5  5.7200
14  60 5  2.3600
15  65 5  2.3600

**Comment by nmueller18** [comment]
Confirmed. Now, `life.table()`only has to be convinced to ignore the column "Age", then this issue can be closed ...

**Comment by nevrome** [comment]
@whamer 362dba4ba02cbc7a6db195f9f490880fa8245104 is [broken](https://travis-ci.org/ISAAKiel/mortAAR/builds/241990526). Please fix and avoid pushing without check.

**Comment by whamer** [comment]
Ich checked it and found the error. My changes corrected the input function; but Martins test environment expects the wrong results (not only Alter istead of Age but also the <NA> Dataframe in the output list).
I don´t know how to play around in the testthat files; so Martin would have to change this. Should I make my changes (Alter -> Age and avoiding <NA> data.frame ) undone?

![grafik](https://user-images.githubusercontent.com/23555619/27037242-9b10d1d6-4f87-11e7-8a55-6e5ba05c65c9.png)

![grafik](https://user-images.githubusercontent.com/23555619/27037257-a83868e2-4f87-11e7-89fa-e8e5339a6b2a.png)

**Comment by MartinHinz** [comment]
Did that with commit 3a4bac1d3bfbf14b889da97839b2600f0e78b612.

**Comment by nevrome** [comment]
With 459dc7bfd666c4769fba00bae80b6f689230510d `life.table()` is now able to deal with the surplus variable "Age". 

```
magdalenenberg %>%
 replace(. == "60-x", "60-70") %>%
 tidyr::separate(a, c("from", "to")) %>%
 dplyr::mutate(from = as.numeric(from), to = as.numeric(to)) %>%
 prep.life.table(
  dec = "Dx", agebeg = "from", ageend = "to", methode = "Standard", age.range = "excluded"
 ) %>%
 life.table()
```

@nmueller18 Please check if the implementation lives up to your expectations. If so, this issue can be closed, I guess?

**Comment by nevrome** [comment]
@nmueller18 Again the question: Does life.table() behave now as expected? I guess not - there still seems to be a problem with the variable "Age".

**Comment by nmueller18** [comment]
Yes, it is still there.

**Comment by nevrome** [comment]
I'm well aware of that - I added it. Please take a look at the implementation yourself. I also changed the documentation to include "Age". Right now it's treated as some kind of alternative for the variable "x". In my opinion that makes sense.

Why does "Age" appear in the output of `prep.life.table()` but should be dropped afterwards? If it's just supposed to work as a visual age class marker for the user, then it would be much better to remove it and instead copy the implementation of "x" also to `prep.life.table()`. What is the **purpose** of "Age"?

**Comment by nmueller18** [comment]
Let's talk about that later this afternoon.

**Comment by MartinHinz** [comment]
@nmueller18 , @nevrome , is the issue now resolved given that you talked yesterday in the afternoon? If so, please close it.

**Comment by nmueller18** [comment]
No, it is not yet. @whamer , @FFaupel , we wanted to change "Age" to "x", if I remembered correctly, and implement the code of the analytical function for that.

**Comment by nevrome** [comment]
I would like to wait until this is done before I adjust life.table()

**Comment by FFaupel** [comment]
I will do it on monday...

**Comment by nevrome** [comment]
`life.table()` should be prepared for your changes! 8acb0b8a0d1cd06734284a1f3fc9e2326b17e3e1
